Reorganize framework hacking panel

ptomato opened the pull request:
https://phabricator.endlessm.com/T23977

ptomato committed:
framework: Add widget for popup menu with choices

This takes the existing popup menu for the logos, and refactors it into
a separate component which is available for all toolboxes. You can
construct it with a map of string IDs to property values, as well as the
widget class to create to represent each choice (e.g. Gtk.Label or
Gtk.Image), the property to set on each widget, and extra properties to
give each widget.

We will use this component for more stuff in the next commit.

https://phabricator.endlessm.com/T23977

ptomato committed:
framework: Reorganize hacking panel

This implements the newest wireframes.

There are two components in the level 2 panel, "Text Cipher" and
"Hyperlinks", which are present in the UI file but hidden since they are
not yet implemented in the model.

https://phabricator.endlessm.com/T23977

cosimoc commented on src/framework/level1.js:
```diff
@@ -34,6 +28,7 @@ var FrameworkLevel1 = GObject.registerClass({
         model.accent_color = this._accentColorButton.rgba;
         model.info_color = this._infoColorButton.rgba;
         model.logo_color = this._logoColorButton.rgba;
+        model.logo_graphic = 'dinosaur';
```
Should this not use `this._logoGroup.value`?

ptomato replied:
Good point.

cosimoc replied:
Still not fixed though?

ptomato replied:
It is fixed, maybe you are looking at an old commit?

cosimoc replied:
Ah, you fixed it in the next commit.

cosimoc commented on src/framework/level1.js:
```diff
@@ -2,30 +2,24 @@
     _init(props = {}) {
         super._init(props);
 
+        const logoChoices = {};
```
I wonder if it wouldn't be easier to instead have this as a `Map`. That way, if you stored it in the menu instead of separately storing the keys and values, you could potentially avoid the double scan in `_updateForPropValue()`

ptomato replied:
Fixed in `popupMenu`.

ptomato commented:
The "hyperlinks" and "cipher" controls are still coming, but here's the existing commits revised.

cosimoc commented on src/popupMenu.js:
```diff
@@ -0,0 +1,92 @@
+
+    _updateForPropValue(propValue) {
+        this._selected[this._itemProp] = propValue;
+        this.value = this._propMap.get(propValue);
```
This line will also cause the one above to be executed as part of the call to `_updateForEnumValue()` through the setter, so the line above could be removed.

cosimoc commented on src/popupMenu.js:
```diff
@@ -0,0 +1,92 @@
+    },
+}, class PopupMenu extends GObject.Object {
+    _init(button, groupDef, ItemClass, itemProp, itemExtraProps, props = {}) {
+        const entries = Object.entries(groupDef);
```
I wonder why you don't pass directly a Map to this method, instead of unpacking `groupDef`

ptomato replied:
In most cases the caller is going to want to define groupDef as a literal. There are no Map literals in JS, only Object literals.

cosimoc commented on src/popupMenu.js:
```diff
@@ -0,0 +1,92 @@
+
+        this._menu.add(this._choices);
+        entries.forEach(([, propValue]) => {
+            const choiceProps = {[itemProp]: propValue, visible: true};
```
One way to eliminate the need for the `propMap` back-table could be to store the enum value as well on this.

ptomato replied:
Done.

ptomato committed:
framework: Implement "Text Cipher" slider

This adds a Caesar cipher to the custom card and banner modules, and a
corresponding model property in the toolbox, and unhides the slider in
the toolbox.

The framework modules also have a "decodefunc" property, this is not
used currently but will be in the Hackdex.

https://phabricator.endlessm.com/T23977

cosimoc commented on src/framework/customModules/banner/hackableDynamic.js:
```diff
@@ -86,8 +94,51 @@ var HackableDynamic = new Module.Class({
     _processText(text) {
         const decomposedText = text.normalize('NFKD');
-        return TextFilters[this.textfilter](decomposedText);
+        let output = '';
```
I'd like to see this factored out to an utility function, since it's used in a few different places

ptomato committed:
framework: Implement "Hyperlinks" switch

This adds a switch with which you can turn off navigation in the app.
Previously navigation would simply be turned off if you had click
sounds, and on in all other cases. This gives you more fine-grained
control to do what you want.

In the code view, it is allowed to write `yes`, `no`, `on`, or `off` as
well as the normal `true` or `false` for the hyperlinks variable. These
alternative boolean constants are unfortunately not yet highlighted.

https://phabricator.endlessm.com/T23977

cosimoc commented on src/framework/customModules/banner/hackableDynamic.js:
```diff
@@ -36,6 +37,14 @@ var HackableDynamic = new Module.Class({
+        decodefunc: GObject.ParamSpec.string('decodefunc', 'Decode Function',
+            'Source code of function to evaluate to decode the text',
+            GObject.ParamFlags.WRITABLE | GObject.ParamFlags.CONSTRUCT,
+            ''),
```
I commented on Slack that this would feel like a good use case for an interface implemented by all the modules that rotate text, so that the boilerplate does not have to be repeated over and over.
But that can be added later in a follow up cleanup, so I won't block the PR on it.
